fish-rust/src/tinyexpr.rs

```diff
@@ -0,0 +1,714 @@
+    ("sqrt"L, Function::Fn1(f64::sqrt)),
+    ("tan"L, Function::Fn1(f64::tan)),
+    ("tanh"L, Function::Fn1(f64::tanh)),
+    ("tau"L, Function::Constant(TAU)),
+];
+
+assert_sorted_by_name!(BUILTINS, 0);
+
+fn find_builtin(name: &wstr) -> Option<Function> {
+    let idx = BUILTINS
+        .binary_search_by_key(&name, |(name, _expr)| name)
+        .ok()?;
+
+    Some(BUILTINS[idx].1)
+}
+
+impl<'s> State<'s> {
+    pub fn new(input: &'s wstr) -> Self {
+        let mut state = Self {
+            start: input,
+            pos: 0,
+            current: Token::End,
+            error: None,
+        };
+        state.next_token();
+        state
+    }
+
+    pub fn error(&self) -> Result<(), Error> {
+        if let Token::End = self.current {
+            Ok(())
+        } else if let Some(error) = self.error {
+            Err(error)
+        } else {
+            // If we're not at the end but there's no error, then that means we have a
+            // superfluous token that we have no idea what to do with.
+            Err(Error {
+                kind: ErrorKind::TooManyArgs,
+                position: self.pos,
+                len: 0,
+            })
+        }
+    }
+
+    pub fn eval(&mut self) -> f64 {
+        return self.expr();
+    }
+
+    fn set_error(&mut self, kind: ErrorKind, pos_len: Option<(usize, usize)>) {
+        self.current = Token::Error;
+        let (position, len) = pos_len.unwrap_or((self.pos, 0));
+        self.error = Some(Error {
+            kind,
+            position,
+            len,
+        });
+    }
+
+    fn no_specific_error(&self) -> bool {
+        !matches!(self.current, Token::Error)
+            || matches!(
+                self.error,
+                Some(Error {
+                    kind: ErrorKind::Unknown,
+                    ..
+                })
+            )
+    }
+
+    /// Tries to get the next token from the input. If the input does not contain enough data for
+    /// another token, `None` is returned. Otherwise, the numbre of consumed characters is returned
+    /// along with either the token, or `None` in case of ignored (whitespace) input.
+    fn get_token(&mut self) -> Option<(usize, Option<Token>)> {
+        debug_assert!(!matches!(self.current, Token::Error));
+
+        let next = &self.start.as_char_slice().get(self.pos..)?;
+
+        // Try reading a number.
+        if matches!(next.first(), Some('0'..='9') | Some('.')) {
+            let mut consumed = 0;
+            let num = wcstod_underscores(*next, get_numeric_locale().decimal_point, &mut consumed)
+                .unwrap();
+            Some((consumed, Some(Token::Number(num))))
+        } else {
+            // Look for a function call.
+            // But not when it's an "x" followed by whitespace
+            // - that's the alternative multiplication operator.
+            if next.first()?.is_ascii_lowercase()
+                && !(*next.first()? == 'x' && next.get(1)?.is_whitespace())
```

Comment:
I'm not sure if the question marks do what we want here.
Given `next.first().is_none()`, the C++ version would go into the else branch ("Look for an operator or special character") where it would report a MissingOperator error.

Reply:
Given `!*next_`, the C++ version would already bail with `type_ = TOK_END;` earlier. You are right about the `next.get(1)?` though; in C++ that would be a null byte that fails `isspace` rather than a bail. My solution isn't the prettiest, but [`Option::is_some_and`](rust-lang/rust#93050) will come soon.